### PR TITLE
perf(ce): single-shot dx_y gradient correction

### PR DESCRIPTION
**Pull request by justinhh4**
## Summary

This PR isolates one optimization on top of the already-merged CE improvements in #1266 (`exp2`) and
#1267 (dtype-aware `num_warps`): replace the per-element true-class gradient selection with a single
post-loop `dx_y` correction.

At the reference shape (`BT=8192`, `V=128256`, full forward + backward), the isolated impact is:

| GPU | dtype | baseline (ms) | this PR (ms) | speedup |
|:----|:------|--------------:|-------------:|--------:|
| B200 | bf16 | 1.4229 | 1.1787 | **1.207x** |
| B200 | fp32 | 1.8820 | 1.8022 | **1.044x** |
| H100 | bf16 | 2.3162 | 2.2734 | **1.019x** |
| H100 | fp32 | 4.5362 | 4.5674 | **0.993x** |

## What changed

The CE gradient subtracts 1 only at the true-class index `y`. Previously, both pass-2 paths evaluated
a per-element `tl.where(X_offsets != y, ...)` across the entire vocabulary. This PR stores the plain
softmax gradient in the loop, then applies one scalar correction at `y` after the loop.

This removes an `O(V)` compare/select used to update one element and replaces it with an `O(1)` store.
A `tl.debug_barrier()` orders the loop's in-place writes before the correction reads and updates
`dx_y`. The correction retains the same weighting, reduction, label-smoothing, and softcap factors.

## Speed benchmarks

`baseline` is `main` with #1266 and #1267 but without this PR's single-shot `dx_y` change.
`speedup = baseline / this PR`, so values above `1.00x` mean this PR is faster.

Each latency is the average of two independent medians. B200 runs were collected on swapped NVIDIA
B200 GPUs; H100 runs were collected on one NVIDIA H100 80GB HBM3. Each median uses 40 timed iterations
after 20 warmup iterations with CUDA events. `full` is forward plus `loss.backward()` with
`grad_output=1.0`, matching the standard last-layer CE benchmark used in #1290.

### bf16

**Speed - full (forward + backward) - vocab sweep (`BT=8192`)**

| vocab | B200 baseline (ms) | B200 this PR (ms) | B200 speedup | H100 baseline (ms) | H100 this PR (ms) | H100 speedup |
|------:|-------------------:|------------------:|-------------:|-------------------:|------------------:|-------------:|
| 32000 | 0.5179 | 0.4455 | **1.163x** | 0.8484 | 0.7890 | **1.075x** |
| 50304 | 0.7967 | 0.6663 | **1.196x** | 1.2223 | 1.1375 | **1.075x** |
| 102400 | 1.3693 | 1.1129 | **1.230x** | 2.0434 | 1.8907 | **1.081x** |
| 128256 | 1.4240 | 1.1802 | **1.207x** | 2.3179 | 2.2790 | **1.017x** |
| 152064 | 1.7677 | 1.4415 | **1.226x** | 2.9131 | 2.7976 | **1.041x** |
| 201088 | 2.4596 | 1.9703 | **1.248x** | 3.8283 | 3.7046 | **1.033x** |
| 262144 | 2.8544 | 2.3144 | **1.233x** | 4.7663 | 4.6687 | **1.021x** |

**Speed - full - `BT` sweep (`V=128256`)**

| BT | B200 baseline (ms) | B200 this PR (ms) | B200 speedup | H100 baseline (ms) | H100 this PR (ms) | H100 speedup |
|---:|-------------------:|------------------:|-------------:|-------------------:|------------------:|-------------:|
| 1024 | 0.3744 | 0.3356 | **1.116x** | 0.5599 | 0.5552 | **1.008x** |
| 2048 | 0.5264 | 0.4595 | **1.146x** | 0.8055 | 0.7913 | **1.018x** |
| 4096 | 0.8250 | 0.6987 | **1.181x** | 1.3031 | 1.2690 | **1.027x** |
| 8192 | 1.4229 | 1.1787 | **1.207x** | 2.3162 | 2.2734 | **1.019x** |
| 16384 | 2.6574 | 2.1671 | **1.226x** | 4.2325 | 4.1427 | **1.022x** |
| 32768 | 5.0804 | 4.1005 | **1.239x** | 8.0368 | 7.8853 | **1.019x** |
| 65536 | 9.9058 | 7.9347 | **1.248x** | 15.6748 | 15.3520 | **1.021x** |

### fp32

**Speed - full (forward + backward) - vocab sweep (`BT=8192`)**

| vocab | B200 baseline (ms) | B200 this PR (ms) | B200 speedup | H100 baseline (ms) | H100 this PR (ms) | H100 speedup |
|------:|-------------------:|------------------:|-------------:|-------------------:|------------------:|-------------:|
| 32000 | 0.6876 | 0.6391 | **1.076x** | 1.0472 | 1.0402 | **1.007x** |
| 50304 | 1.0058 | 0.9423 | **1.067x** | 1.5734 | 1.6267 | **0.967x** |
| 102400 | 1.7260 | 1.6429 | **1.051x** | 3.7195 | 3.7151 | **1.001x** |
| 128256 | 1.8806 | 1.8026 | **1.043x** | 4.5767 | 4.5742 | **1.001x** |
| 152064 | 2.3784 | 2.2591 | **1.053x** | 5.3767 | 5.3722 | **1.001x** |
| 201088 | 3.2668 | 3.1830 | **1.026x** | 7.0200 | 7.0008 | **1.003x** |
| 262144 | 4.0601 | 3.9960 | **1.016x** | 9.0001 | 9.0043 | **1.000x** |

**Speed - full - `BT` sweep (`V=128256`)**

| BT | B200 baseline (ms) | B200 this PR (ms) | B200 speedup | H100 baseline (ms) | H100 this PR (ms) | H100 speedup |
|---:|-------------------:|------------------:|-------------:|-------------------:|------------------:|-------------:|
| 1024 | 0.4218 | 0.4141 | **1.019x** | 0.8335 | 0.8518 | **0.979x** |
| 2048 | 0.6379 | 0.6167 | **1.034x** | 1.3555 | 1.3693 | **0.990x** |
| 4096 | 1.0574 | 1.0149 | **1.042x** | 2.4686 | 2.4647 | **1.002x** |
| 8192 | 1.8820 | 1.8022 | **1.044x** | 4.5362 | 4.5674 | **0.993x** |
| 16384 | 3.5883 | 3.4314 | **1.046x** | 8.6929 | 8.7241 | **0.996x** |
| 32768 | 6.8904 | 6.5826 | **1.047x** | 17.0411 | 17.0401 | **1.000x** |
| 65536 | 13.6248 | 13.0628 | **1.043x** | 33.7193 | 33.6499 | **1.002x** |

## Correctness

The full CE suite passes across bf16 and fp32 configurations, including class weights, softcap,
label smoothing, z-loss, and `ignore_index`.

### CODE REVIEW

**Comment by vaibhavjindal** [comment]
@justinhh4 since #1266 and #1267 are merged, we can try merging this now. Can you update the PR description reporting just the benefits from single shot dx_y improvement? This will help us evaluate the individual impact of this change.

**Comment by justinhh4** [comment]
@vaibhavjindal thanks for the suggestions. I have updated the PR body, please take a look!

File: src/liger_kernel/ops/cross_entropy.py [hunk 1]
```diff
@@ -216,8 +216,7 @@ def liger_cross_entropy_kernel(
                 X_block += 2 * lse_square_scale * lse * X_block
                 # smoothing term
                 X_block += -eps
-                # special handle dx_y
-                X_block = tl.where(X_offsets != y, X_block, X_block - (1 - label_smoothing))
+                # dx_y corrected once after the loop (removed the per-element tl.where)
```
Comment (vaibhavjindal) [inline src/liger_kernel/ops/cross_entropy.py]: nit: Remove the "(removed the per-element tl.where)" comment

File: src/liger_kernel/ops/cross_entropy.py [hunk 2]
```diff
@@ -226,8 +225,7 @@ def liger_cross_entropy_kernel(
                 softmax_X = tl.exp2((X_block - m) * LOG2_E) / d
                 # derivative of original_loss
                 dloss_ori = (1 - label_smoothing) * softmax_X
-                # specially handle dx_y
-                dloss_ori = tl.where(X_offsets != y, dloss_ori, dloss_ori - (1 - label_smoothing))
+                # dx_y corrected once after the loop (removed the per-element tl.where)
```
Comment (vaibhavjindal) [inline src/liger_kernel/ops/cross_entropy.py]: ditto

**Review by vaibhavjindal — COMMENTED** [review]
LGTM, great catch. Just one minor nit about removing some comments.

**Review by vaibhavjindal — APPROVED** [review]
